------------------------------------------------------------
revno: 389
fixes bug(s): https://launchpad.net/bugs/726938
committer: Scott Moser <smoser@ubuntu.com>
branch nick: trunk
timestamp: Mon 2011-03-21 21:35:54 -0400
message:
  fix bug with resizefs module
  
  instead of using blkid on /dev/root, create a device node ourselves with
  the correct device number of '/', then use blkid and resize2fs on that.
  
  I believe the problem was that /dev/root was occasionally not being present
  due to race.
------------------------------------------------------------
revno: 388
fixes bug(s): https://launchpad.net/bugs/739694
committer: Scott Moser <smoser@ubuntu.com>
branch nick: trunk
timestamp: Mon 2011-03-21 16:24:52 -0400
message:
  fix bug in part-handler code, that broke working part-handlers (LP: #739694)
------------------------------------------------------------
revno: 387
committer: Scott Moser <smoser@ubuntu.com>
branch nick: trunk
timestamp: Thu 2011-03-10 09:36:28 -0500
message:
  mention sources.list.tmpl in sources.list.tmpl so user knows what to change
------------------------------------------------------------
revno: 386
committer: Scott Moser <smoser@ubuntu.com>
branch nick: trunk
timestamp: Tue 2011-03-08 10:52:11 -0500
message:
  fix issue where apt-update was not done unless explicitly set
=== modified file 'ChangeLog'
--- old/ChangeLog	2011-02-18 23:44:25 +0000
+++ new/ChangeLog	2011-03-22 01:35:44 +0000
@@ -1,3 +1,10 @@
+0.6.2:
+ - fix bug where update was not done unless update was explicitly set.
+   It would not be run if 'upgrade' or packages were set to be installed
+ - fix bug in part-handler code, that prevented working part-handlers
+   (LP: #739694)
+ - fix bug in resizefs cloud-config that would cause trace based on
+   failure of 'blkid /dev/root' (LP: #726938)
 0.6.1:
  - fix bug in fixing permission on /var/log/cloud-init.log (LP: #704509)
  - improve comment strings in rsyslog file tools/21-cloudinit.conf

=== modified file 'cloudinit/CloudConfig/cc_apt_update_upgrade.py'
--- old/cloudinit/CloudConfig/cc_apt_update_upgrade.py	2011-03-04 01:04:24 +0000
+++ new/cloudinit/CloudConfig/cc_apt_update_upgrade.py	2011-03-08 15:49:28 +0000
@@ -55,8 +55,10 @@
             log.error("Failed to run debconf-set-selections")
             log.debug(traceback.format_exc())
 
+    pkglist = util.get_cfg_option_list_or_str(cfg,'packages',[])
+
     errors = [ ]
-    if update:
+    if update or len(pkglist) or upgrade:
         try:
             cc.update_package_sources()
         except subprocess.CalledProcessError as e:
@@ -72,7 +74,6 @@
             log.debug(traceback.format_exc())
             errors.append(e)
 
-    pkglist = util.get_cfg_option_list_or_str(cfg,'packages',[])
     if len(pkglist):
         try:
             cc.install_packages(pkglist)

=== modified file 'cloudinit/CloudConfig/cc_resizefs.py'
--- old/cloudinit/CloudConfig/cc_resizefs.py	2011-03-03 22:15:12 +0000
+++ new/cloudinit/CloudConfig/cc_resizefs.py	2011-03-22 01:23:18 +0000
@@ -18,6 +18,9 @@
 import cloudinit.util as util
 import subprocess
 import traceback
+import os
+import stat
+import tempfile
 
 def handle(name,cfg,cloud,log,args):
     if len(args) != 0:
@@ -29,21 +32,38 @@
 
     if not resize_root: return
 
-    log.debug("resizing root filesystem on first boot")
+    # this really only uses the filename from mktemp, then we mknod into it
+    (fd, devpth) = tempfile.mkstemp()
+    os.unlink(devpth)
+    os.close(fd)
+    
+    try:
+       st_dev=os.stat("/").st_dev
+       dev=os.makedev(os.major(st_dev),os.minor(st_dev))
+       os.mknod(devpth, 0400 | stat.S_IFBLK, dev)
+    except:
+        log.warn("Failed to make device node to resize /")
+        raise
 
-    cmd = ['blkid', '-c', '/dev/null', '-sTYPE', '-ovalue', '/dev/root']
+    cmd = [ 'blkid', '-c', '/dev/null', '-sTYPE', '-ovalue', devpth ]
     try:
         (fstype,err) = util.subp(cmd)
     except subprocess.CalledProcessError as e:
-        log.warn("Failed to get filesystem type via %s" % cmd)
+        log.warn("Failed to get filesystem type of maj=%s, min=%s via: %s" %
+            (os.major(st_dev), os.minor(st_dev), cmd))
         log.warn("output=%s\nerror=%s\n", e.output[0], e.output[1])
+        os.unlink(devpth)
         raise
 
+    log.debug("resizing root filesystem (type=%s, maj=%i, min=%i)" % 
+        (fstype.rstrip("\n"), os.major(st_dev), os.minor(st_dev)))
+
     if fstype.startswith("ext"):
-        resize_cmd = [ 'resize2fs', '/dev/root' ]
+        resize_cmd = [ 'resize2fs', devpth ]
     elif fstype == "xfs":
-        resize_cmd = [ 'xfs_growfs', '/dev/root' ]
+        resize_cmd = [ 'xfs_growfs', devpth ]
     else:
+        os.unlink(devpth)
         log.debug("not resizing unknown filesystem %s" % fstype)
         return
 
@@ -52,5 +72,8 @@
     except subprocess.CalledProcessError as e:
         log.warn("Failed to resize filesystem (%s)" % resize_cmd)
         log.warn("output=%s\nerror=%s\n", e.output[0], e.output[1])
+        os.unlink(devpth)
         raise
-        
+
+    os.unlink(devpth)
+    return

=== modified file 'cloudinit/__init__.py'
--- old/cloudinit/__init__.py	2011-02-07 19:36:33 +0000
+++ new/cloudinit/__init__.py	2011-03-21 20:18:27 +0000
@@ -363,7 +363,7 @@
         self.handlercount=self.handlercount+1
 
         # write content to instance's handlerdir
-        handlerdir = self.get_ipath("handler")
+        handlerdir = self.get_ipath("handlers")
         modname  = 'part-handler-%03d' % self.handlercount
         modfname = modname + ".py"
         util.write_file("%s/%s" % (handlerdir,modfname), payload, 0600)

=== modified file 'templates/sources.list.tmpl'
--- old/templates/sources.list.tmpl	2010-08-31 14:24:58 +0000
+++ new/templates/sources.list.tmpl	2011-03-10 14:36:14 +0000
@@ -4,6 +4,7 @@
 \## a.) add 'apt_preserve_sources_list: true' to /etc/cloud/cloud.cfg
 \##     or do the same in user-data
 \## b.) add sources in /etc/apt/sources.list.d
+\## c.) make changes to template file /etc/cloud/templates/sources.list.tmpl
 \###
 
 # See http://help.ubuntu.com/community/UpgradeNotes for how to upgrade to

